app/src/main/kotlin/com/mapbox/vision/teaser/MainActivity.kt

```diff
@@ -1,55 +1,531 @@
+                        animate().translationY(0f)
+                                .setDuration(500L)
+                                .setInterpolator(AccelerateDecelerateInterpolator())
+                                .start()
+                    }
+                } else {
+                    speed_limit_next.hide()
+                }
+            }
+        }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
+        super.onCreate(savedInstanceState)
+
+        if (!SystemInfoUtils.isVisionSupported()) {
+            AlertDialog.Builder(this)
+                    .setTitle(R.string.vision_not_supported_title)
+                    .setView(
+                            TextView(this).apply {
+                                setPadding(dpToPx(20f).toInt())
+                                movementMethod = LinkMovementMethod.getInstance()
+                                isClickable = true
+                                text = HtmlCompat.fromHtml(
+                                        getString(R.string.vision_not_supported_message),
+                                        HtmlCompat.FROM_HTML_MODE_LEGACY
+                                )
+                            }
+                    )
+                    .setCancelable(false)
+                    .show()
+
+            VisionLogger.e(
+                    "BoardNotSupported",
+                    "System Info: [${SystemInfoUtils.obtainSystemInfo()}]"
+            )
+        }
+
+        setContentView(R.layout.activity_main)
+
+        if (!allPermissionsGranted() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            requestPermissions(getRequiredPermissions(), PERMISSIONS_REQUEST_CODE)
+        } else {
+            onPermissionsGranted()
+        }
+
+        soundsPlayer = SoundsPlayer(this)
+    }
+
+    private fun onPermissionsGranted() {
+        isPermissionsGranted = true
+
+        signSize = dpToPx(64f).toInt()
+        lineHeight = dpToPx(40f).toInt()
+        margin = dpToPx(8f).toInt()
+
+        back.setOnClickListener { onBackClick() }
+        segm_container.setOnClickListener { setVisionMode(VisionMode.Segmentation) }
+        sign_detection_container.setOnClickListener { setVisionMode(VisionMode.Classification) }
+        det_container.setOnClickListener { setVisionMode(VisionMode.Detection) }
+        distance_container.setOnClickListener { setVisionMode(VisionMode.Safety) }
+        safety_mode_container.hide()
+        line_detection_container.setOnClickListener { setVisionMode(VisionMode.Lanes) }
+
+        initRootLongTap()
+        fps_performance_view.hide()
+
+        initArNavigationButton()
+        initReplayModeButton()
+        tryToInitVisionManager()
+    }
+
+    private fun initRootLongTap() {
+        root.setOnLongClickListener {
+            if (fps_performance_view.visibility == View.GONE) {
+                fps_performance_view.show()
```

Comment:
how about small extension `reverseVisibility` ?

Reply:
Wow, great idea!

Reply:
small note:
there are 3 states for visibility.

Maybe add smth like that
`reverseVisibility(invisibleState = View.Invisible)` and when it's really necessary set `View.Gone`

Reply:
I think it can have the following semantics : 
reverse visible <-> gone, do NOT touch invisible

Reply:
Or maybe it's worth to have two different functions, 
`toggleVisibleGone`
`toggleVisibleInvisible`

Reply:
I'd prefer to do it after all modifications later, added issue with notes
https://github.com/mapbox/mapbox-vision-android-teaser/issues/191